api/v1beta1/csiaddonsjob_types.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2021 The Kubernetes-CSI-Addons Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TargetSpec defines the targets on which the operation can be
+// performed.
+type TargetSpec struct {
+	// PVC represents PersistentVolumeClaim name.
+	PVC string `json:"pvc,omitempty"`
+
+	// PVC represents PersistentVolume name.
+	PV string `json:"pv,omitempty"`
+}
+
+// ReclaimSpaceSpec defines the required parameters for ReclaimSpace
+// operation.
+type ReclaimSpaceSpec struct {
+	// Target represents targets on which the operation can be
+	// performed.
+	// +kubebuilder:validation:Required
+	Target TargetSpec `json:"target"`
+
+	// Parameters is a key-value map with driver configurations for
+	// performing the operation.
+	// +kubebuilder:validation:Optional
+	Parameters map[string]string `json:"parameters,omitempty"`
+}
+
+// OperationSpec defines the desired operation of CSIAdoonsJobSpec.
+type OperationSpec struct {
+	// ReclaimSpace repesents ReclaimSpace operation.
+	ReclaimSpace ReclaimSpaceSpec `json:"reclaimSpace,omitempty"`
+}
+
+// OperationState defines the state of the operation.
+type OperationState string
+
+const (
+	// Succeeded represents the Succeeded operation state.
+	Succeeded OperationState = "Succeeded"
+
+	// Failed represents the Failed operation state.
+	Failed OperationState = "Failed"
+)
+
+// CSIAddonsJobSpec defines the desired state of CSIAddonsJob
+type CSIAddonsJobSpec struct {
+	// Operation represents the operation to be performed on the volume.
+	// Current Supported operation is "ReclaimSapce".
```

Comment:
Why are we hard-coding support for add-ons into the `csi-addons` spec here? Is the intent for `csi-addons` not to be an extensible interface? The interface is no longer very extensible or independent once we start hard-coding add-on implementations into the add-on spec. I think this is anti-pattern to the intent of the CSI ecosystem and falls back into an easy but poorly-designed system.

Reply:
The different operations in the spec follow the same design as [Volumes in a Pod](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#volume-v1-core). Each Volume has a single `<type>VolumeSource` attribute, everything is strongly typed there.

Not only validation of the Spec is easier when it is strongly typed. An other important advantage is that users can easily see what is possible, without having to figure out what parameters can be used for a specific `string` formatted operation.

Using a CRD per operation would be my 2nd choice, but I wonder if that makes things less user-friendly (a scheduled version like `CronJob` is needed in addition to the `Job`). The number of CRDs might grow a lot, extending the singe CRD with a new operation might be easier for users?

Reply:
I think this assumes that the design for Pod volumes is (1) the correct analog to use and (2) that the pod volume design doesn't have flaws we should consider. I have been in COSI design meetings with one of the people involved with creating the PVC and PV design, and he often expresses regret over some of the design decisions made there. It is good reminder to review whether prior art that exists in Kubernetes is actually "good" and whether a better design might meet needs better and allow for more flexibility.

In my estimation, this design would be more closely approximated the design of StorageClasses and not pod volumes. Even considering pod volumes as inspiration, I really don't think the pod volume API is above reproach. For the benefit that exists from strong typing, I also see a lot of the provider-specific stuff in that API as technical debt and not a feature. We are seeing the kubernetes community move away from provider-specific APIs with CSI in order to use the more generic CSI interface. 

I still think having a generic and non-prescriptive interface creates a better design because it does not create tight coupling between the CSI addons API and arbitrary implementations of csi addons. I think a design with looser coupling would be a more preferable one for pod volumes were kuberenetes ever to do a v2 API, and I am pretty certain that the storage sig would push to change that API in v2.

And also to be sure, the design of StorageClasses is not above reproach. String-string maps of arbitrary parameters can be a little challenging to use. Based on seeing the trends in Kubernetes of moving away from tightly coupled APIs (CSI for a large example), I would encourage us to have generic API interfaces that are not tied directly to specific implementations of CSI addons but that still have an amount of specificity that can help developers.

Reply:
To be a little more transparent about some of the thoughts driving my considerations, I think that if we want the CSI addons API to be used and accepted by organizations outside of Red Hat, we would be better off having generic API components that follow the trend of StorageClasses more than the trend of Pod volumes. This will make it easier for outside organizations to consume the API without having to do work to push code changes to this API in addition to implementing their own addon.

If these addons are useful for people but we make the addons hard to develop and consume for those outside the Kubernetes ecosystem, it is more likely that a Kubernetes special interest group (SIG) will crop up to fill the gap, and we may have to pivot to match what the Kubernetes community is doing in the future. Or it's possible that CSI addons aren't interesting to other organizations, and we are the only ones who want to do this, and so it may not be a big issue. I truly don't know.

All that said, I think that organizations within Red Hat will still benefit from having a decoupled design, as it should mean addons can be created a little more quickly and without requiring as much back-and-forth between different engineering teams. But ultimately, Red Hat has to cooperate with Red Hat to meet corporate deliverables, and so that is not as big of a concern.